pkg/staking/stakingcontract/contract.go

```diff
@@ -0,0 +1,198 @@
+	overlayNonce           []byte
+}
+
+func New(
+	owner common.Address,
+	stakingContractAddress common.Address,
+	bzzTokenAddress common.Address,
+	transactionService transaction.Service,
+	nonce []byte,
+) Contract {
+	return &contract{
+		owner:                  owner,
+		stakingContractAddress: stakingContractAddress,
+		bzzTokenAddress:        bzzTokenAddress,
+		transactionService:     transactionService,
+		overlayNonce:           nonce,
+	}
+}
+
+func (s *contract) sendTransaction(ctx context.Context, callData []byte, desc string) (*types.Receipt, error) {
+	request := &transaction.TxRequest{
+		To:          &s.stakingContractAddress,
+		Data:        callData,
+		GasPrice:    sctx.GetGasPrice(ctx),
+		GasLimit:    1600000,
+		Value:       big.NewInt(0),
+		Description: desc,
+	}
+
+	txHash, err := s.transactionService.Send(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.transactionService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if receipt.Status == 0 {
+		return nil, transaction.ErrTransactionReverted
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendDepositStakeTransaction(ctx context.Context, owner common.Address, stakedAmount big.Int, nonce common.Hash) (*types.Receipt, error) {
+
+	callData, err := stakingABI.Pack("depositStake", owner, &stakedAmount, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.sendTransaction(ctx, callData, depositStakeDescription)
+	if err != nil {
+		return nil, fmt.Errorf("deposit stake: stakedAmount %d: %w", &stakedAmount, err)
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendGetStakeTransaction(ctx context.Context, overlay []byte) (*big.Int, error) {
+
+	callData, err := stakingABI.Pack("stakeOfOverlay", overlay)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := s.transactionService.Call(ctx, &transaction.TxRequest{
+		To:   &s.stakingContractAddress,
+		Data: callData,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("get stake: overlayAddress %d: %w", overlay, err)
+	}
+
+	results, err := erc20ABI.Unpack("stakeOfOverlay", result)
+	if err != nil {
+		return nil, err
+	}
+	return abi.ConvertType(results[0], new(big.Int)).(*big.Int), nil
+}
+
+func (s *contract) DepositStake(ctx context.Context, stakedAmount big.Int, overlay swarm.Address) error {
+	prevStakedAmount, err := s.GetStake(ctx, overlay)
+	if err != nil {
+		return err
+	}
+
+	if len(prevStakedAmount.Bits()) == 0 {
+		if stakedAmount.Cmp(MinimumStakeAmount) == -1 {
+			return ErrInsufficentStakeAmount
+		}
+	}
+
+	balance, err := s.getBalance(ctx)
+	if err != nil {
+		return err
+	}
+
+	if balance.Cmp(&stakedAmount) < 0 {
+		return ErrInsufficientFunds
+	}
+
+	_, err = s.sendDepositStakeTransaction(ctx, s.owner, stakedAmount, common.BytesToHash(s.overlayNonce))
+	if err != nil {
+		return err
+	}
+	//TODO: verify if we need receipt as well as service for staking
+	//TODO: logic for receipt would be added here
```

Comment:
`DepositStake` currently does wait receipt - this method calls `sendDepositStakeTransaction`, which calls `sendTransaction` which calls `WaitForReceipt`.

Please, consider removing this TODO.

Reply:
umm so this todo was left in case there was some code op that i missed which needed the receipt, ill remove it after my sync session with Esad.

Reply:
Alright, generally please consider not checking in any `TODO`s in code because it should always get resolved before code is merged. In case there is technical dept there should be opened issue because that's only way technical depth is tracked.

In this manner please consider removing other `TODO`s from the code (there are some in the beginning of this file)